rtengine/ahd_demosaic_RT.cc

```diff
@@ -0,0 +1,229 @@
+
+#ifdef _OPENMP
+#pragma omp parallel
+#endif
+{
+    int progresscounter = 0;
+    float (*pix), (*rix)[3];
+    float ldiff[2][4], abdiff[2][4];
+    float (*lix)[3];
+    char *buffer = (char *) malloc (12 * TS * TS * sizeof(float) + 2 * TS * TS * sizeof(uint16_t)); /* 1053 kB per core */
+    float (*rgb)[TS][TS][3]  = (float(*)[TS][TS][3]) buffer;
+    float (*lab)[TS][TS][3]  = (float(*)[TS][TS][3])(buffer + 6 * TS * TS * sizeof(float));
+    uint16_t (*homo)[TS][TS] = (uint16_t(*)[TS][TS])(buffer + 12 * TS * TS * sizeof(float));
+
+#ifdef _OPENMP
+    #pragma omp for collapse(2) schedule(dynamic) nowait
+#endif
+    for (int top = 2; top < height - 5; top += TS - 6) {
+        for (int left = 2; left < width - 5; left += TS - 6) {
+            //  Interpolate green horizontally and vertically:
+            for (int row = top; row < top + TS && row < height - 2; row++) {
+                for (int col = left + (FC(row, left) & 1); col < std::min(left + TS, width - 2); col += 2) {
+                    pix = &(rawData[row][col]);
+                    float val0 = 0.25f * ((pix[-1] + pix[0] + pix[1]) * 2
+                                  - pix[-2] - pix[2]) ;
+                    rgb[0][row - top][col - left][1] = median(val0, pix[-1], pix[1]);
+                    float val1 = 0.25f * ((pix[-width] + pix[0] + pix[width]) * 2
+                                  - pix[-2 * width] - pix[2 * width]) ;
+                    rgb[1][row - top][col - left][1] = median(val1, pix[-width], pix[width]);
+                }
+            }
+
+            //  Interpolate red and blue, and convert to CIELab:
+            for (int d = 0; d < 2; d++)
+                for (int row = top + 1; row < top + TS - 1 && row < height - 3; row++) {
+                    int cng = FC(row + 1, FC(row + 1, 0) & 1);
+                    for (int col = left + 1; col < std::min(left + TS - 1, width - 3); col++) {
+                        pix = &(rawData[row][col]);
+                        rix = &rgb[d][row - top][col - left];
+                        lix = &lab[d][row - top][col - left];
+                        if (FC(row, col) == 1) {
+                            rix[0][2 - cng] = CLIP(pix[0] + (0.5f * (pix[-1] + pix[1]
+                                                       - rix[-1][1] - rix[1][1] ) ));
+                            rix[0][cng] = CLIP(pix[0] + (0.5f * (pix[-width] + pix[width]
+                                                       - rix[-TS][1] - rix[TS][1])));
+                            rix[0][1] = pix[0];
+                        } else {
+                            rix[0][cng] = CLIP(rix[0][1] + (0.25f * (pix[-width - 1] + pix[-width + 1]
+                                                        + pix[+width - 1] + pix[+width + 1]
+                                                        - rix[-TS - 1][1] - rix[-TS + 1][1]
+                                                        - rix[+TS - 1][1] - rix[+TS + 1][1])));
```

Comment:
Didn't check this (and the above) relative indexing dance. I assume, you've double-checked it...

Reply:
@Floessie Do you mean the indexing by cng (ColourNonGreen)?